libs/community/langchain_community/tools/gaode/weather.py

```diff
@@ -0,0 +1,73 @@
+"""
+Tool implementations for the gaode (https://www.amap.com) weather API.
+
+Documentation: https://lbs.amap.com/api/webservice/summary
+API keys:      https://console.amap.com/dev/index
+"""
+
+import json
+import os
+from typing import Any, Type
+
+import requests
+from langchain_core.pydantic_v1 import BaseModel, Field
+from langchain_core.tools import BaseTool
+
+CITY_CODE_URL = "https://restapi.amap.com/v3/config/district"
+WEATHER_URL = "https://restapi.amap.com/v3/weather/weatherInfo"
+
+
+class GaodeWeatherInput(BaseModel):
+    """Input for the gaode weather tool."""
+
+    city: str = Field(description="city that require weather query")
+
+
+class GaodeWeatherTool(BaseTool):
+    """Gaode weather query tool"""
+
+    name = "gaode_weather_tool"
+    description = "Query for weather information based on the input city"
+    args_schema: Type[BaseModel] = GaodeWeatherInput
+    return_direct = True
+
+    def _run(self, *args: Any, **kwargs: Any) -> str:
+        city = kwargs.get("city")
+        if city is None:
+            return "Invalid city input!"
+
+        api_key = os.getenv("GAODE_API_KEY")
+        if api_key is None:
+            return "Please apply for gaode api_key first"
```

Comment:
Should we add an API key to the class attributes and raise an error if the tool is instantiated without one, instead?

Reply:
OK, I've optimized my implementation.